Readme.md
![module info](https://nodei.co/npm/es.shim.png?downloads=true&downloadRank=true&stars=true)

# es.shim
> `es.shim` is an extend module for letting you can use some future api in current Node.js version.
> Also some useful api for you.


+ Obejct
    * empty()


+ Obejct.prototype
    * merge()


+ Array.prototype
    * includes()



+ Date
    * isDate()


+ Date.prototype
    * getFullWeek()
    * getWeek()
    * format()


+ String.prototype
    * splice()
    * htmlspecialchars()
    * tohtml()
    * xss()
    * escape()
    * padStart()
    * padEnd()



+ global
    * gmdate()
    * empty()
    


## Usage

### 1. Object.prototype.merge()
```javascript
let obj1 = {a: 123, b: 456}
let obj2 = {a: 22, c: 44}
let obj3 = {c: 11, e: 55}

o1.merge(o2)
// now obj1 is {a: 22, b: 456, c: 44}
// nothing to obj2 

o1.merge(o2, o3)
// obj1 will be {a: 22, b: 456, c: 11, e: 55}
// nothing to obj2 & obj3

```

### 2. Object.empty()
```javascript
Object.empty({}) // true
Object.empty({a: 213}) // false
Object.empty([]) // true
Object.empty([null]) // false
Object.empty([undefined]) // false

```


### 3. Array.prototype.includes()
```javascript
let arr = [1, '3', 54, 32, 'foo']

arr.includes(1) // true
arr.includes(3) // false
arr.includes('bar') // false
arr.includes('54') // false

```


### 4. Date.isDate()
```javascript

Date.isDate(new Date()) // true
Date.isDate({}) // false
Date.isDate(['bar']) // false
Date.isDate('foo') // false

```


### 5. Date.prototype.format(format)
> `format` can be these below:
> - Y (with century) eg. 1970,2017
> - y (without century) eg. 70, 117
> - m month, 01-12
> - n month, 1-12
> - d date, 01-31
> - j date, 1-31
> - H hours 00-23
> - h hours 00-12
> - G hours 0-23
> - g hours 0-12
> - i minutes, 00-59
> - s seconds, 00-59
> - W how many weeks from 01-01 this year
> - w how many weeks from 01 this month
> - D week name, like Mon, Tue, Wed, Thu, Fri, Sat, Sun

```javascript

new Date().format() // default 2017-02-08 12:11:23
new Date().format('Y-m-d') // 2017-02-08
new Date().format('Y/n/j') // 2017/2/8
new Date().format('Y年n月j日 第W周') // 2017年2月10日 第6周
new Date('Wed Feb 10 2016 23:34:04 GMT+0800 (CST)').format() // 2016-02-10 23:34:04
new Date('2016-08-10T13:14:44.000Z').format() //2016-08-10 21:14:44
new Date(1470834884000).format('') //2016-08-10 21:14:44

```


### 6. String.prototype.splice(start, len[, fill])
- start `<Integer>`
- len `<Integer>`
- fill `<String>`

```javascript
let str = 'Hello baby';

str.splice(0, 5) // return  ' baby'
console.log(str) // nothing tostr, so it return 'Hello baby'

str.splice(0, 5, 'Love') //return 'Love baby'

str.splice(6, 0, 'world, ')// return 'Hello world, baby'

str = str.splice(6) //return  'Holle '

```


### 7. String.prototype.htmlspecialchars([sign])
> Just like php's function - `htmlspecialchars`
- sign `<String>` (ENT_QUOTES/ENT_NOQUOTES)

```javascript
let str = `<script>alert('hello world')</script>`

str.htmlspecialchars() // &lt;script&gt;alert('hello world')&lt;/script&gt;
str.htmlspecialchars('ENT_QUOTES') // &lt;script&gt;alert(&#39;hello world&#39;)&lt;/script&gt;

```


### 8. String.prototype.tohtml()
```javascript
let str = `&lt;script&gt;alert(&#39;hello world&#39;)&lt;/script&gt;`
str.tohtml() // <script>alert('hello world')</script>

```




### 9.global.empty()

```javascript
empty('') //true
empty(null) //true
empty(undefined) //true
empty([]) //true
empty({}) //true
empty({a: 123}) //false
empty(0) //true
empty('0') //false
empty(1) //false

```


### 10.global.gmdate()
> Base on Date.prototype.format

```javascript
gmdate() // 2017-02-08 12:11:23
gmdate('Y-m-d') // 2017-02-08
gmdate('', 1470834884000) //2016-08-10 21:14:44

```


History.md
1.1.0 / 2018-06-29
==================
  * add Number.parse


1.0.1 / 2018-05-25
==================
  * delete global extends
  * add Promise.defer extend


0.0.3 / 2017-03-17
==================
  * Optimise String extend function.

0.0.1 / 2017-02-28
==================
  * Initial release

lib/string.js
/**
 * 
 * @authors yutent (yutent@doui.cc)
 * @date    2017-02-27 22:01:10
 *
 */

"use strict";


//类似于Array 的splice方法
if(!String.prototype.splice){
    Object.defineProperty(String.prototype,
        'splice',
        {
            value: function(start, len, fill){
                let length = this.length,
                    argLen = arguments.length;

                fill = fill === undefined ? '' : fill

                if(argLen < 1){
                    return this
                }

                //处理负数
                if(start < 0){
                    if(Math.abs(start) >= length)
                        start = 0
                    else
                        start = length + start
                }

                if(argLen === 1){
                    return this.slice(0, start)
                }else{
                    len -= 0;

                    let strl = this.slice(0, start),
                        strr = this.slice(start + len);

                    return strl + fill + strr
                }
            },
            enumerable: false
        })
}



//同php的htmlspecialchars函数
if(!String.prototype.htmlspecialchars){
    Object.defineProperty(String.prototype,
        'htmlspecialchars',
        {
            value: function(sign){
                let str = this.replace(/&(?!\w+;)/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')

                if(sign === 'ENT_QUOTES')
                    return str.replace(/"/g, '&quot;').replace(/'/g, '&#39;')
                else if(sign === 'ENT_NOQUOTES')
                    return str
                else
                    return str.replace(/"/g, '&quot;')
            },
            enumerable: false
        })
}


//htmlspecialchars的还原
if(!String.prototype.tohtml){
    Object.defineProperty(String.prototype,
        'tohtml',
        {
            value: function(){
                return this.replace(/&lt;/ig, '<')
                        .replace(/&gt;/ig, '>')
                        .replace(/&quot;/ig, '"')
                        .replace(/&#39;/g, '\'')
                        .replace(/&amp;/ig, '&')
            },
            enumerable: false
        })
}


//简单的过滤xss
if(!String.prototype.xss){
    Object.defineProperty(String.prototype,
        'xss',
        {
            value: function(){
                let str = this.htmlspecialchars('ENT_QUOTES')
                    str = str.replace(/(document\.cookie)|(document\.write)|(\.parentNode)|(window\.location)|(\.innerHTML)/g, '')
                        .replace(/(%0[0-8bcef])|(%1[0-9a-f])/g, '')
                    return str
            },
            enumerable: false
        })
}

// js特殊字符的转义
if(!String.prototype.escape){
    Object.defineProperty(String.prototype,
        'escape',
        {
            value: function(){
                return this.replace(/('|"|&|\\|\}|\{|\(|\)|;|=|\,|&)/g, '\\$1')
            },
            enumerable: false
        })
}





// padStart & padEnd
if(!String.prototype.padStart){
    Object.defineProperty(String.prototype,
        'padStart',
        {
            value: function(len, fill){
                let alen = arguments.length,
                    length = this.length,
                    ilen = len - length;

                if(alen < 1 || ilen < 1)
                    return this

                if(alen < 2 || typeof fill !== 'string')
                    fill = ' '

                while(fill.length < ilen){
                    fill += fill
                }

                return fill.slice(0, ilen) + this
            },
            enumerable: false
        })

    Object.defineProperty(String.prototype,
        'padEnd',
        {
            value: function(len, fill){
                let alen = arguments.length,
                    length = this.length,
                    ilen = len - length;

                if(alen < 1 || ilen < 1)
                    return this

                if(alen < 2 || typeof fill !== 'string')
                    fill = ' '

                while(fill.length < ilen){
                    fill += fill
                }

                return this + fill.slice(0, ilen)
            },
            enumerable: false
        })
}